### PR TITLE
implement user sign up endpoint (task4)

### PR DESCRIPTION
**Pull request by haitham-awad**

### CODE REVIEW

**Comment by haitham-awad** [comment]
[Task 4: Implement User Sign-Up Endpoint (API) Use FastAPI if Python](https://trello.com/c/JgFUGld9/5-task-4-implement-user-sign-up-endpoint-api-use-fastapi-if-python)

File: backend/main.py [hunk 1]
```diff
@@ -1,6 +1,32 @@
-from fastapi import FastAPI
+from fastapi import FastAPI, status, HTTPException
+from pydantic import BaseModel
+import bcrypt
+
+#User object to save user info
+class User(BaseModel):
+    email: str
+    password: str
+    username: str
```
Comment (jerkur22) [inline backend/main.py]: should be in schema folder 
schemas are for structure of data that the API wants and returns

File: backend/main.py [hunk 2]
```diff
@@ -1,6 +1,32 @@
...
+    email_set.add(user.email)
+
+    #encrypt password and save to database
+    encrypted_password = bcrypt.hashpw(user.password.encode('utf-8'), bcrypt.gensalt())
+    database[user.email] = {
+        "email":user.email,
+        "password":encrypted_password,
+        "username":user.username
+    }
+    return { "message": "User registered" }
```
Comment (jerkur22) [inline backend/main.py]: move to routers folder 
routers/user.py
```python
from fastapi import APIRouter

router = APIRouter()

@router.post("/api/register")
async def create_user_profile():
    # your code
    return { "message": "User registered" }

File: backend/main.py [hunk 3]
```diff
@@ -1,6 +1,32 @@
...
+
```
Comment (jerkur22) [inline backend/main.py:8]: instead put following here 
```python
from routers import user  # import the user router

app = FastAPI()

# Include the router with a prefix (optional)
app.include_router(user.router)``

**Review by jerkur22 — CHANGES_REQUESTED** [review]
looks good just needs to move around
Put comment somewhere everywhere you add code detailing what task it is with comment

File: backend/main.py (hunk 1 above)
Comment (arikamat) [reply backend/main.py:9]: lets put this in models folder instead. Lets reserve schemas for db sqlalchemy things if needed in future

File: tests/backend/test_create_user.py
Comment (arikamat) [inline tests/backend/test_create_user.py:1]: fix file name as per contributing/styling standards

File: backend/main.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 from fastapi import FastAPI
+from backend.routers.user import router  # import the user router
```
Comment (arikamat) [inline backend/main.py]: In backend/routers/__init__.py: put in ``from .user import router as UserRouter``
Thecn change this import ``from backend.routers import UserRouter``

File: tests/backend/test_create_user.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+from fastapi.testclient import TestClient
+from fastapi import FastAPI
+import pytest
+from backend.main import app
```
Comment (arikamat) [inline tests/backend/test_create_user.py]: in backend's __init__.py have a 
``from .main import app``
Then change this import to ``from backend import app``

**Review by arikamat — APPROVED** [review]
Great job

LGTM!

**Review by jerkur22 — APPROVED** [review]
looks good

**Pull request merged**